src/VisualStudio/Core/Def/LanguageService/AbstractLanguageService`2.cs

```diff
@@ -71,9 +73,22 @@ internal abstract partial class AbstractLanguageService<TPackage, TLanguageServi
     /// </remarks>
     private bool _isSetUp;
 
+    protected abstract string ContentTypeName { get; }
+    protected abstract string LanguageName { get; }
+    protected abstract string RoslynLanguageName { get; }
+    protected abstract Guid DebuggerLanguageId { get; }
+
     protected AbstractLanguageService(TPackage package)
     {
         Package = package;
+
+        Debug.Assert(!this.Package.JoinableTaskFactory.Context.IsOnMainThread, "Language service should be instantiated on background thread");
+
+        this.EditorOptionsService = this.Package.ComponentModel.GetService<EditorOptionsService>();
+        this.Workspace = this.Package.ComponentModel.GetService<VisualStudioWorkspaceImpl>();
+        this.EditorAdaptersFactoryService = this.Package.ComponentModel.GetService<IVsEditorAdaptersFactoryService>();
+        this.AnalyzerFileWatcherService = this.Package.ComponentModel.GetService<AnalyzerFileWatcherService>();
+        this._languageDebugInfo = CreateLanguageDebugInfo();
```

Comment:
this could all move inside constructor, as we know this constructor is always instantiated on BG thread.  makes things much simpler/cleaner too.